### ISSUE TITLE
Collections Specification

### ISSUE THREAD

**Issue by DragaDoncila** [ome/ngff#31]
**What is an image collection?**
A collection of images is a semantic grouping of two or more associated ome-ngff images and/or image-labels. 

This definition could include
- Images which do not share a physical coordinate space e.g. training dataset of images containing bees
- Images which share a physical coordinate space and whose storage specification must support sufficient metadata to determine this positioning e.g. high-content screening plates and wells
- A hierarchy of image groups of arbitrary depth which may or may not share physical coordinates
- Other things…?

**What workflows should it support?**
The specification should support implementations being able to traverse the image collection and, where relevant, map the associated metadata to the physical coordinate space for loading these images.

Ideally, the specification should provide sufficient information at each level of a hierarchical grouping to allow for the loading of both the entire collection, and the loading of an arbitrary level of the hierarchy. This can be important when wanting to share/view partial datasets or update only small parts of the entire collection.

Where labels or other related data is provided (e.g. meshes, points…), the specification should support being able to associate any member of the image collection with its associated labels, regardless of the level in the hierarchy.

The OME-NGFF spec is close to supporting this functionality with the HCS specification which allows the positioning of wells into rows and plates. The main drawbacks of this specification are
- It is too specific to be easily used for images which ARE physically associated but are not HCS acquisitions
- It may be difficult to understand for researchers who are not working with HCS images but nevertheless wish to store their collection in OME-NGFF format
- It does not support an arbitrary depth of groupings
- It does not support collections which are not physically associated

**What should it be called?**
- Dataset - this term is already used in various places so may not be the best choice
- Collection - a general enough term which is currently mostly unused
- Hierarchical definition - there is a case for this specification being a hierarchy of specifications, with each one defining a more tightly bound collection e.g.
    - Bag - associated images with no metadata
    - Stack - associated images which overlap in physical space
    - Panorama - associated images which stitch together in physical space

Ideally, the names used in the base specification would be general enough to support a broad variety of use cases and tailored use cases could be demonstrated using examples in the documentation.

**Reference specifications**
[BDV XML Files](https://imagej.net/BigDataViewer.html#Multi-view_data_converted_to_XML.2FHDF5)
[SVG](https://www.w3.org/TR/SVG2/coords.html)
[TrakEM2](https://imagej.net/TrakEM2)
[Napari Plugin for image-label collections](https://github.com/DragaDoncila/napari-compressed-labels-io)
[mobie grid view of many sources](https://github.com/mobie/mobie/issues/20)

**Related**
[Image.SC discussion on collections](https://forum.image.sc/t/ome-ngff-combine-multiple-images-into-one-dataset/47619/7)
[Live notes from latest community call](https://forum.image.sc/t/next-call-on-next-gen-bioimaging-data-tools-feb-23/48386/9)
[HCS Specification](https://ngff.openmicroscopy.org/latest/#hcs-layout)

**What next?**
I think we should first decide on whether we want to support arbitrary levels in the hierarchy and whether we want a general spec which we can “inherit” from for more detailed specs, or whether we want one spec to rule them all.

My vote is that we define the most generic collection (a “bag” of images) which works with arbitrary levels of grouping (it’s collections all the way down), and then work to add to it for more complex collections. I will be working on this over the coming week and will post here once I have something working, but of course would love to hear what everyone’s thoughts are on the best way forward.

**Comment by tischi** [comment]
@DragaDoncila Thank you very much for the detailed post! It makes a lot of sense and I am looking forward to whatever you come up with! Incidentally we (cc @constantinpape ) were also working on this topic during the past few days.

I also ping @d-v-b

I would like to add a notion and would be curious to hear opinions: 

Currently I think that I would prefer storing images within a zarr container without any hierarchy, i.e. just as a flat list. The main reason is simplicity for the reader and writer libraries (the current HCS specifications does not follow this). Anything that imposes a hierarchy would be handled by the collections specification, which I think could be seen as metadata that specifies how to display and layout several images together.

**Comment by constantinpape** [comment]
> Currently I think that I would prefer storing images within a zarr container without any hierarchy, i.e. just as a flat list.

This means you would like to store images as 2d arrays and volumes as 3d arrays, correct?
I created #35 to discuss this.

**Comment by tischi** [comment]
> This means you would like to store images as 2d arrays and volumes as 3d arrays, correct?

I was not gonna enter the 3D vs 5D discussion here, but just wanted to say that I feel that structuring the zarr like this: https://ngff.openmicroscopy.org/latest/#hcs-layout feels overly complex to me.

**Comment by constantinpape** [comment]
> I was not gonna enter the 3D vs 5D discussion here, but just wanted to say that I feel that structuring the zarr like this: https://ngff.openmicroscopy.org/latest/#hcs-layout feels overly complex to me.

Ok, so your point is to rather have a flat hierarchy of images in the zarr container:
```
image1/
image2/
image3/
image4/
...
```
and then define the potential hierarchies in the collections metadata (just a mock-up):
```json
{
  "well1": ["image1", "image2"],
  "well2": ["image3", "image4"]
}
```

**Comment by tischi** [comment]
Yes, exactly.

The way I see it conceptually is that a multi-well plate is a specific _layout_ of a _bag of images_ and, as such, should be covered by our _collections specification_, which I would currently see as metadata that exists _independent_ of the way we store the raw image data. What do you think?

**Comment by will-moore** [comment]
One feature that the current HCS layout gives us is a URL to a specific Well. So I can open a specific Well like: https://hms-dbmi.github.io/vizarr/v0.1?source=https://s3.embassy.ebi.ac.uk/idr/zarr/v0.1/plates/2551.zarr/A/1
Demo movie at https://twitter.com/will_j_moore/status/1322187662762090497

I guess you could try to use a URL `?query` or a `#fragment` to refer to a Well or other subgroup. E.g. `path/to/plate.zarr/#A1`

**Comment by tischi** [comment]
> a URL to a specific Well

I see that this is cool, but I am afraid that (i) these hierarchies make it harder parse an ome.zarr and (ii) it is not flexible; for example, I guess I cannot produce an single URL to show me all the images that were subjected to the same biological treatment (which may be several wells).

**Comment by DragaDoncila** [comment]
> The way I see it conceptually is that a multi-well plate is a specific layout of a bag of images and, as such, should be covered by our collections specification, which I would currently see as metadata that exists independent of the way we store the raw image data.

I completely agree that the metadata and storage should be independent, because I think this also provides the opportunity to support a wider range of custom metadata. For example this:

```
{
  "well1": ["image1", "image2"],
  "well2": ["image3", "image4"]
}
```

could easily be this (for some geographical feature learning model):
```
{
  "lakes": ["image1", "image2"],
  "mountains": ["image3", "image4"]
}
```
I guess that's what I was thinking of when I said 
> tailored use cases could be demonstrated using examples in the documentation.

I like the idea of a flat set of images with the hierarchy determined entirely by the metadata. That certainly seems the easiest way to support an arbitrary level of hierarchy without ending up with a very complex storage structure.

**Comment by joshmoore** [comment]
> https://github.com/ome/ngff/issues/31#issuecomment-787820475 Currently I think that I would prefer storing images within a zarr container without any hierarchy, i.e. just as a flat list.

Is this a MAY or a MUST? And what happens when/if someone does make use of the folder structure available in Zarr/N5/HDF5?

**Comment by will-moore** [comment]
Re @tischi "flexibility and biological treatment" - I'm wondering if there must be a single 'hierarchy' in the container, e.g. If we can have multiple. E.g.

```
{
  "well1": ["image1", "image2"],
  "well2": ["image3", "image4"]
}
```
And:
```
{
  "aquisition1": ["image1", "image3"],
  "aquisition2": ["image2", "image4"]
}
```
or 
```
{
  "drug1": ["image1", "image3"],
  "drug2": ["image2", "image4"]
}
```
Those are all different ways to grouping the images.
But if you have:
```
{
  "well1": ["image1", "image2"],
  "well2": ["image3", "image4"],
  "drug1": ["image1", "image3"],
  "drug2": ["image2", "image4"]
}
how do you know which groups are mutually exclusive. E.g. which ones are Wells vv Treatments?

Having multiple hierarchies might provide more flexibility, but this makes it harder to understand how to view the data.
Instead, it might make more sense to only have a single hierarchy (like a file-system) and then add other metadata in other ways?

**Comment by tischi** [comment]
@will-moore 

My current idea would be to have **no** hierarchy on the data storage level, but provide the possibility to specify different "views" on the data on the metadata level. Something along the lines:

```
views: 
{
   "well_based": {...},
   "treatment_based":{...}
}
default_view: "well_based"
```

Does that make sense to you?

**Comment by tischi** [comment]
> Is this a MAY or a MUST? And what happens when/if someone does make use of the folder structure available in Zarr/N5/HDF5?

Personally, I'd be for a MUST, i.e. not support hierarchies and then ignore anything stored at deeper levels.
But, obviously, that's just my personal opinion. Very curious to hear other opinions!

**Comment by constantinpape** [comment]
> Personally, I'd be for a MUST, i.e. not support hierarchies and then ignore anything stored at deeper levels.
> But, obviously, that's just my personal opinion. Very curious to hear other opinions!

I think I am not such a big fan of the MUST here. There are some use cases where hierarchies make a lot of sense to keep the data ordered. As an simple example:
I have segmentations computed with two different algorithms and two hyperparameters for the algorithms, and I want to store them in the same container to compare them with some viewer that can ingest it.
For this use case having 
```
algorithm1/
    parameter_set1
    parameter_set2
    parameter_set3
algorithm2/
    parameter_set1
    parameter_set2
```
is a more natural (and easier to navigate) way of storing this then
```
algorithm1_parameter_set1
algorithm1_parameter_set2
algorithm1_parameter_set3
algorithm2_parameter_set1
algorithm2_parameter_set1
```

**Comment by tischi** [comment]
> There are some use cases where hierarchies make a lot of sense to keep the data ordered

OK, fair enough :)

I guess the question is whether, in practice, one would navigate the data via the "views" or via the folder structure. If one changes ones mind at some point about the folder structure, this could be quite expensive in terms of reordering all the data (at least that's how I understood how the object stores work), while it would be very cheap to just replace the views, isn't it?

**Comment by constantinpape** [comment]
> I guess the question is whether, in practice, one would navigate the data via the "views" or via the folder structure. If one changes ones mind at some point about the folder structure, this could be quite expensive in terms of reordering all the data (at least that's how I understood how the object stores work), while it would be very cheap to just replace the views

Sure, reordering the folder structure is not such a good idea but also not necessary because we can have multiple views for the same data. But having a hierarchical folder structure does not change anything about the views except that there will be some `\` in the data names.

**Comment by tischi** [comment]
> But having a hierarchical folder structure does not change anything about the views except that there will be some \ in the data names.

Yes, that is true. I guess it'd be fine with a MAY, but should we then maybe "strongly encourage" that there is a default_view specified that one could go to in order to efficiently find out what's in the dataset, without having to go through the whole "folder structure"? (I am also thinking about our experience that things like `cd` and `ls` sometimes are super slow on object stores).

**Comment by will-moore** [comment]
I'm also a little hazy on object stores, but my impression is that all the 'paths' within a bucket are really just 'keys'. So I imagine they could be changed without moving the data on disk. 
So, as @constantinpape's said, I'm not sure there's really much difference between `algorithm1/parameter_set1` and `algorithm1_parameter_set1`. I don't think you can browse to `algorithm1/` on an object store.
This is why you need to specify all the paths to child objects in the group metadata.

However, since you won't always be working with object stores, allowing `algorithm1/parameter_set1` could let you browse the data via `algorithm1/` elsewhere, so I think this could be helpful. Also conceptually helpful to tokenise the path in this way. So I think we should allow `/` in the path names.

E.g. this could be valid:
```
{
  "lakes": ["day1/image1", "image2"],
  "mountains": ["image3", "day2/image4"]
}
```
Any reason not to allow this?

**Comment by tischi** [comment]
@will-moore I think that's fine and a very good point. On a file system there are some benefits to this and on an object store there are no disadvantages.

**Comment by will-moore** [comment]
OK, so it looks like there's enough consensus here to start on something a bit more concrete.
Aiming for something that is just a list of images in its simplest form, but can include more metadata without a breaking change. 

# Option 1 
A `path/to/collection/` directory would include a `.zattrs` file that defines a "collection" because it MUST include the `collection` key, which MUST contain an `images` list:

```
{
  "collection": {
    "images": [
      {"path": "image1"},
      {"path": "dir_1/image2"},
    ]
  }
}
```
Each `path` is a path/to/directory containing an `OME-Zarr` image.
Each item in the `images` list MUST have a `path`, but MAY also have other attributes (TBD: e.g. `id`, `name`, `timestamp`, etc. Maybe even `'row': 0, 'column': 1`, for a grid layout.) We should probably not allow any user-chosen key-value data here, since that could lead to breaking changes if we add keys to the spec. So maybe a `properties: {}` for user-defined metadata.

# Option 2
An alternative is to use the "path" as the ID/key of each image. Any reason not to do this? (for [labels-metadata](https://ngff.openmicroscopy.org/latest/#label-md) we decided not to use an ID as key because the ID was a number which is not a valid key in JSON.). This protects from having 2 identical 'path' values which could be possible above.
```
{
  "collection": {
    "images": {
      "image1": {},    # empty if we don't have any other info
      "dir_1/image2": {"row": 0, "column": 1},
      "dir_2/image3": {"properties": {"rating": 5}},
    }
  }
}
```

# Other optional metadata
Within the `collection`, alongside `images` we could imagine other metadata such as layout:
```
  "layout": {
    "type": "grid",   # or 'auto-grid'
    "rows": [
      {"name": "A"}, {"name": "B"}, {"name": "C"}
    ],
    "columns": [
      {"name": "1"}, {"name": "2"}, {"name": "3"}
    ],
  },
```
and groupings. I guess we could use the path as the identifier of each image.
```
  "groups": {
    "lakes": ["image1", "dir_1/image2"],
    "mountains": ["dir_2/image3"]
  }
```
Which could mean that the "images" list/dict above is not needed (if we don't have any other metadata, and every image is in a group)? BUT it simplifies the spec to say that `images` MUST exist, and it's not hard to always add it.

So, is everyone happy with Option 1 or Option 2? Or would like to suggest improvements to whichever is their favourite?
The other metadata can be decided later, but any suggestions welcome.

**Comment by joshmoore** [comment]
> https://github.com/ome/ngff/issues/31#issuecomment-791273590 So I imagine they could be changed without moving the data on disk.

No. To move objects around in object storage is *always* a copy/delete operation.

> https://github.com/ome/ngff/issues/31#issuecomment-791033216 I guess the question is whether, in practice, one would navigate the data via the "views" or via the folder structure.

It sounds like we're struggling with the semantics of having one "hard-coded" hierarchy beside the additional collections. In the ome-zarr-py implementation (and we could work to formalize this), there's a generator pattern. You start at the group you're given and then ask for what it "points to" and then process that. You will *always* start from a single group, so perhaps we're saying that you will _only_ use the metadata of the given group for the objects that are generated.

**Comment by tischi** [comment]
> It sounds like we're struggling with the semantics of having one "hard-coded" hierarchy beside the additional collections.

@joshmoore I agree. Maybe, for simplicity, we could restrict this issue to discussion of the additional collections? and make an extra issue but "hard-coded" hierarchy?

> A path/to/collection/ directory would include a .zattrs file 

@will-moore Do I get it right that currently one such .zattrs file would contain only one collection? Meaning that to specify multiple additional collection we would need several `path/to/collection/ .zattrs`? I guess that's fine, but then I guess somewhere there should be information how to find them?

> So, is everyone happy with Option 1 or Option 2?

Option 2 looks more concise, so maybe slight preference for that one.

In terms of the layout, instead of specifying row and column, I think specifying a translation in physical coordinates may also be an option.

```
{
  "collection": {
    "images": {
      "image1": {"translate": [0,0,0], "name": "A"},   
      "dir_1/image2": {"translate": [10,0,0], "name": "B"},
      "dir_2/image3": {"translate": [20,0,0], "name": "C"},
    }
  }
}
```

**Comment by constantinpape** [comment]
I also prefer option2. And as @tischi brought up I think it's important to think about how to map different collections for the same data (or subsets of it), either in the same .zattrs or distributed into different ones in some defined pattern.

**Comment by will-moore** [comment]
I was thinking the multiple `groups` above were different groupings of images in a collection. But I guess that's not enough, e.g. if each image has a different `translate` or other property in each collection.

So the simplest way to support multiple `collections` is to make the dict -> list, and plural:

```
# .zattrs
{
  "collections": [
    {
      "name": "first collection",
      "images": {
        "image1": {"translate": [0,0,0], "name": "A"},   
        "dir_1/image2": {"translate": [10,0,0], "name": "B"},
        "dir_2/image3": {"translate": [20,0,0], "name": "C"},
      }
    },
    {
      "images": {
        "dir_2/image3": {},
        "dir_2/image4": {},
      }
    }
  ]
}
```

**Comment by DragaDoncila** [comment]
>  Do I get it right that currently one such .zattrs file would contain only one collection? Meaning that to specify multiple additional collection we would need several path/to/collection/ .zattrs? I guess that's fine, but then I guess somewhere there should be information how to find them?

I think if we want to easily support images being opened both as part of their collection and on their own then it would make sense to have each image as its own well-formed ome-zarr, including a .zattrs file? It would mean either duplication of some metadata, or a top level .zattrs which only contains the necessary information for traversing the collection i.e. the snippet @will-moore posted just above

**Comment by will-moore** [comment]
Yes, in the examples I've posted, there would be a full OME-Zarr in each of the `images` paths. E.g. `dir_1/image2/` would contain `.zattrs` etc.

**Comment by joshmoore** [comment]
Just to clarify one issue that will become more important with Zarr V3, each of those paths contains an `OME-Zarr image`.  In the future, there will be a root file which will define the entire `OME-Zarr fileset`.

**Comment by will-moore** [comment]
I'm going to try and see how we might migrate the metadata in the current HCS spec into this collections spec.
This is based on the current HCS spec: https://ngff.openmicroscopy.org/latest/#plate-md
and also includes proposed changes from https://github.com/ome/ngff/pull/24/ (adds `row_index` and `column_index`).

Here is a 6-well Plate, with 2 acquisitions (only 1 image in the 2nd acquisition).
The only custom info that is not generic to the collections spec is the acquisition `starttime` which is therefore in a user-defined `properties` dictionary. We should probably specify that any custom attributes should go in a `properties` dictionary to avoid clashing with future spec key-words.

cc @melissalinkert

Anyone else want to try using this spec proposal with a sample of their current data needs, to see if it works for you and suggest any other key-words that we might need?

```
# .zattrs

{
  "collections": [
    {
      "name": "HCS plate 01",
      "images": {
        "2020-10-10/A/1": {
          "row_index": 0,
          "column_index": 0
        },
        "2020-10-10_run2/A/1": {
          "row_index": 0,
          "column_index": 0
        },
        "2020-10-10/A/2": {
          "row_index": 0,
          "column_index": 1
        },
        "2020-10-10/A/3": {
          "row_index": 0,
          "column_index": 2
        },
        "2020-10-10/B/1": {
          "row_index": 1,
          "column_index": 0
        },
        "2020-10-10/B/2": {
          "row_index": 1,
          "column_index": 1
        },
        "2020-10-10/B/3": {
          "row_index": 1,
          "column_index": 2
        }
      },
      "layout": {
        "type": "grid",
        "rows": [
          {"name": "A"},
          {"name": "B"},
          {"name": "C"}
        ],
        "columns": [
          {"name": "1"},
          {"name": "2"},
          {"name": "3"}
        ]
      },
      "groups": {
        "acquisition_1": {
          "name": "Meas_01(2012-07-31_10-41-12)",
          "properties": {
            "starttime": 1343731272000,
          },
          "images": [
            "2020-10-10/A/1",
            "2020-10-10/A/2",
            "2020-10-10/A/3",
            "2020-10-10/B/1",
            "2020-10-10/B/2",
            "2020-10-10/B/3"
          ]
        },
        "acquisition_2": {
          "name": "Meas_02(201207-31_11-56-41)",
          "properties": {
            "starttime": 1343735801000,
          },
          "images": [
            "2020-10-10_run2/A/1"
          ]
        }
      }
    }
  ]
}
```

**Comment by will-moore** [comment]
As suggested by @sbesson I checked for any info that can be stored in the current HCS spec but wasn't in the example above. I only found `id` and `maximumfieldcount` for the acquisitions. Initially I felt these could go under each `group` in the user-defined `properties`. But that's really just a place for stuff that's not part of the spec, and since these are part of the current spec, I moved them up to be attributes of each `group` object. But, if it's felt we want to reduce the scope of the collections spec then they could go under `properties`.

Since there haven't been any objections to the latest proposal, can I assume that everyone is happy with it?

If so, my next steps would be to convert an existing HCS OME-Zarr to the format above (which should be possible without moving any images on disk) and then to look at updating the viewing of that data in `ome-zarr-py` (for `napari`) or in `vizarr` (cc @manzt) to see if there are any blockers to the current features.

**Comment by tischi** [comment]
Sorry if I missed it @will-moore: how are sites (aka well sub-positions) treated with your proposal?

**Comment by will-moore** [comment]
@tischi - Apologies I missed your question from 5 months ago. But trying to pick this up again...
I think Well sub-positions could be stored as additional values of each Image. E.g. to use the `PositionX` and `PositionY` from an OME WellSample (which also has `PositionXUnit`, `PositionYUnit`):

```
"images": {
        "2020-10-10/A/1": {
          "row_index": 0,
          "column_index": 0,
          "position_x": 100,
          "position_y": 50,
        },
        "2020-10-10/A/1": {
          "row_index": 0,
          "column_index": 0,
          "position_x": 10,
          "position_y": 5,
        },
```

**Comment by d-v-b** [comment]
why not 
```
...
position: [10, 5],
units: ['um','um'],
axes: [x, y],
...
```

Baking dimension names (which  are actually data) into the keys of the metadata seems brittle.

**Comment by will-moore** [comment]
Thanks for the feedback. 
I can see that a generic solution is generally better than hard-coded dimensions, but if it's too generic then how do clients know how to interpret it? What values to do allow for axes? Anything? or x, y, z, c, t (same as for an image)?
Also, if you apply the same approach to `row_index` and `column_index` then what does your example look like?
You have position `10 um (x)` and `row 0`, so what are your units?

**Comment by d-v-b** [comment]
If I understand this issue correctly (which might not be the case!) this metadata describes a space with named dimensions, and the position of each image in that space. So there's no need to have properties like `position_x` or `column_index` -- you can do something like: 
```json
{
  "images": {
    "2020-10-10/A/1": {
      "position": [
        "A",
        "1"
      ]
    }
  },
  "layout": {
    "type": "grid",
    "dimensions": [
      {
        "rows": [
          {
            "name": "A"
          },
          {
            "name": "B"
          },
          {
            "name": "C"
          }
        ]
      },
      {
        "columns": [
          {
            "name": "1"
          },
          {
            "name": "2"
          },
          {
            "name": "3"
          }
        ]
      }
    ]
  }
}
```
Of course consuming software will need to know how to parse the `layout` object, and it will need to know what a `grid` is, etc. But with this approach there's no spatial /indexing information baked into individual object keys. Images just have a `position`, which is an array of N elements each of which belongs to a set defined by `layout`.

**Comment by will-moore** [comment]
What I believe @tischi was meaning by "well sub-positions" was that for each Well (e.g. A1)has a bunch of Images at different positions within the Well. So each image has a `row` and `column`, but also has a x-position and y-position of e.g. 10 um.

I don't know how you describe that in the examples above.
The 2 examples you have these attributes for an Image
```
position: [10, 5],
units: ['um','um'],
axes: [x, y],
```
and
```
"position": [
        "A",
        "1"
      ]
```

I want to store all these attributes for a single Image, so what does that look like? How do you combine both of those examples?

**Comment by will-moore** [comment]
The OME team discussed this Collections spec the other day. We'd like to start by implementing some of the core components discussed above, and leave aside other parts still to be decided for the time being.

Specifically, a top-level `.zattrs` contains a list of `collections`. Each `collection` has an optional `name` and a `dict` of `images` where each key is the `path/to/image` and the value is a `dict`. This dict could probably contain a `properties` dict for custom data, so as to avoid clashing with any properties that subsequently get added to the `images` spec.

```
# .zattrs

{
  "collections": [
    {
      "name": "Optional Collection name",
      "images": {
        "path/to/image/": {
        # TBD - spec-defined image properties
        "properties": {}        # custom properties
        }
      }
    }
  ]
}
```

Is there anything here that doesn't look right or could be improved before we start to work on this?
Hopefully this provides a good start that can be extended as needed.

**Comment by tischi** [comment]
In [MoBIE](https://mobie.github.io/specs/mobie_spec.html), we currently have it a bit different, something like:

```
"myViewName": {
      "isExclusive": false,
      "sourceDisplays": [
        {
          "imageDisplay": {
            "color": "blue",
            "contrastLimits": [
              0.0,
              1000.0
            ],
            "opacity": 1.0,
            "sources": [
              "path/to/image1/", "path/to/image2/"
            ]
          }
        }
      ],
```

You could ignore most of the above, but I think a point one may consider is that "properties" such as contrast settings can be applied to multiple images. We thought this is convenient, e.g., for a HTM plate view where you typically want to render all images with the same display settings. 

However, I must admit that your version feels easier to digest and thus may be preferable. 

@constantinpape what do you think?

**Comment by glyg** [comment]
> "properties" such as contrast settings can be applied to multiple images

Wouldn't it make sense to allow for a `"properties"` entry in the top level (for each collection) to define defaults, to be overridden by individual image settings?

**Comment by will-moore** [comment]
Like this:
```
{
  "collections": [
    {
      "name": "Optional Collection name",
      "properties": {}.       // custom properties of the collection
      "images": {
        "path/to/image/": {
        "properties": {}        // custom properties of the image
        }
      }
    }
  ]
}
```
That makes sense to me. 👍

**Comment by will-moore** [comment]
You could also add `"properties": {}` to `groups` as in https://github.com/ome/ngff/issues/31#issuecomment-802064440,
but not planning on adding `groups` in the first round, unless this is urgent for anyone?

**Comment by tischi** [comment]
Another thing to consider: how do you want to deal with the label mask images? I think for the current spec this could be captured within the "properties" of an image. For example, there could be a property `showLabelMask`. 

In MoBIE we don't store label masks together with images, but separately. Thus in a above example there is the `imageDisplay`, which could also be a `segmentationDisplay`, if one wanted to show label masks. We made separate fields for those, because, for us, the "properties" of a `segmentationDisplay` were too different from an `imageDisplay` and it was thus more convenient to treat them separately.

**Comment by will-moore** [comment]
@tischi
In your example, if you were to have multiple `segmentationDisplay`s for an `imageDisplay`, it would look like this?

```
"sourceDisplays": [
  {"imageDisplay":{"sources": [...]}},
  {"segmentationDisplay":{"sources": [...]}},
  {"segmentationDisplay":{"sources": [...]}},
```

In the current ngff spec, 1 or more label images can be stored within a "labels" directory underneath the "image" directory.
Could that work for you? Or do you want to use a label image with more than one image?

**Comment by tischi** [comment]
> it would look like this?

Yes, exactly

> Or do you want to use a label image with more than one image?

That's a good question. I can imagine use-cases for this, for example in multi-modal microscopy. Let's say you acquired the same specimen both with LM and EM. And you did the segmentation on the LM image. Then it can be very interesting and relevant to see where those labels end up in the EM image. Thus it could be that one wants to specify a "collection=view" where one only sees the EM and the LM labels. 

Maybe one could do it like this:

- load LM, with property `showLabels` set to `true`, and property `isVisible` set to `false` (where `isVisible` would refer to the LM and not the labels).
- load EM

Like this one could only see the EM and the LM labels.

Some more comments: 

I forgot that there can be multiple label images associated with one image. That means a boolean `showLabels` property would not suffice, but we would have to give the path (which I don't think is a problem).

There could be a segmentation algorithm that, in above example, uses information from both the LM and EM image. Thus one may not want to associate the resulting label mask with just one of them. I think one could store it separately:  https://github.com/ome/ngff/issues/43
I guess, that means we actually do need something in the collection spec to directly just load a label mask?!

**Comment by tischi** [comment]
Another thought: In MoBIE, we always have the logic to do something on multiple sources. For example, take some display settings and apply it to a list of sources. Or, take a list of sources and arrange them in a grid. We thought this can help to make the JSON a bit more concise and also makes it easy to specify to arrange some images in a grid (as needed, e.g. for an HTM plate). I think a disadvantage of this approach is that I don't think that works very well in terms of "interactively saving a collection view". What I mean by this is that a user changes image display settings and maybe also the image location in a viewer and then wants to "save the current view". For this, it is much easier, to simply loop through all open images and serialize their display and transformation settings individually. In fact, the current MoBIE spec also supports this because one _can_ just always specify one source (i.e. a list with one item) per "action". 

Long story short, I am a bit confused whether it could be simpler to have a logic "image with properties" rather than "property with images" (which, as said, however would cover "image with properties" if you put only one item in the list).

**Comment by tischi** [comment]
Looking at this, is it a stupid idea to replace the notion of "properties" with "actions"? Like "assign a color", "transform", ...

**Comment by glyg** [comment]
One of the collection use case I can think of is a collection of renderings for SMLM. In those, one of the properties would be a reference to the rendering method (a DOI to the relevant paper?). Not sure how the "action" keyword would fit here?

**Comment by tischi** [comment]
The action could be "render" on the SMLM data, and then a property of the "render" action would be the "rendering_method".

**Comment by will-moore** [comment]
OK, so maybe in the future we could consider allowing `labels` to exist outside of images.
I think the minimal collections spec we've got here doesn't prevent that.

The notion of "actions" seems more procedural, like a workflow, which we could consider at some point? Whereas "properties" is purely declarative (fixed state). So I'd prefer we start with `properties` for now.

I'm in two minds about the "collections" list I suggested above, thinking that a single `"collection": {}` as first proposed might be preferable.

This way, you can point to a single collection with a URL or path like `/path/to/dir/` and that dir contains `.zattrs` like:

```
# .zattrs
{
  "collection": {
      "name": "Optional Collection name",
      "properties": {}.       // custom properties of the collection
      "images": {
        "path/to/image/": {
          "properties": {}        // custom properties of the image
        }
      }
    }
}
```

You still can't have a URL to a Well or a subset of Images (which is OK, see https://github.com/ome/ngff/issues/31#issuecomment-787871374 above) but at least you have a URL to a collection of Images, instead of a URL to *multiple* collections. The possibility of multiple collections makes it harder for a client viewer to handle that URL.

Thinking about vizarr or napari here, but I guess the same would apply to other viewers.
We can later add e.g. "groups" to "map different collections for the same data (or subsets of it)" as requested at https://github.com/ome/ngff/issues/31#issuecomment-792249255 instead of a top-level `collections` list.

When dealing with large numbers of images (too many for 1 collection) we may want to provide hierarchy via a "collection" containing other `"collections": ["path/to/child/collection", "another_child_collection"]`.

Apologies for changing even this minimal proposal. This still good?

**Comment by glyg** [comment]
> When dealing with large numbers of images (too many for 1 collection) we may want to provide hierarchy via a "collection" containing other `"collections": ["path/to/child/collection", "another_child_collection"]`.

Why should the `"collections" be a list of collections and not a dict as "images"` -- it seems to me the recursion is more natural if `"images"` and `"collections"` have the same type (one could also extend to "tables" this way)?

**Comment by tischi** [comment]
In the end it's just a word (properties vs actions), put I think your point about this potentially being more "procedural" is interesting. With your current proposal, how would you envision applying 2 different transformations after each other in a defined order to one of the images in the collection?

I very much like the idea that a collection can be represented by a URL such that it can easily be opened! So maybe adding a group for each collection could be a good idea.

Also I agree that collections of collections can be useful. In fact, for HTM, all "wells" could be their own collection and then the "plate" could be a collection of all the "wells". Then you could have an URL for each well. What do you think?

**Comment by will-moore** [comment]
Yes, good point @glyg . `"collections": {"path/to/child/collection":{"name": "foo"},  "another_child_collection":{}}` makes sense.

**Comment by sbesson** [comment]
> one could also extend to "tables" this way

This raises a question I have been pondering about while reading the thread above. How tight to the `image` concept should be this specification, or at least its initial version? Assuming there will be a short-term need to register other data types than image in the lowest level collection, should the top-level key be more agnostic e.g. `datasets` as in the `multiscales` specification for instance?  Or is it better to be specific and have additional top-level keys for other data types?

**Comment by will-moore** [comment]
Just to re-iterate: I would only suggest we use `"properties":{}` on an image, group or collection to hold custom key-value pairs that are not yet part of the spec. But the whole idea is that this data should be migrated to form part of the spec ASAP, so that different clients know what they mean.

So you could do something like:
```
"images": {
  "image1": {
    "transformations": [
      {"affine": "values"}   // etc
      {"affine": "values"}   // etc?!
    ],
    "properties": {}    // custom data
}
```
But that puts `"transformations"` in the "collections" spec.
If that transformation applies to a single image, you probably want to store it underneath the image `.zattrs`, alongside or within the "multiscales" metadata.
If the transformations are a relationship between 2 images then maybe something different, but there is a Transformation spec issue at https://github.com/ome/ngff/issues/28 so I guess it will be discussed there.

**Comment by will-moore** [comment]
@sbesson I think there was a preference for "collection" over "dataset" 
"this term is already used in various places so may not be the best choice".
I don't have a strong preference but I don't think that "collection" excludes any types of data.
I could imagine:

```
"collection": {
  "images": {},
  "labels": {},
}
```
Not sure what is meant by `extend to "tables"`?

**Comment by sbesson** [comment]
Agreed, I am absolutely not proposing that the top-level collection should be called `datasets` as I also think this term is overloaded.

Ignoring `tables` for a moment and taking your example of a collection of images an labels, two specialized Zarr groups, I can see 2 options: either have a top-level key where the data type is enforced:

```
"collection": {
  "images": {}, # must be images
  "labels": {}, # must be labels
}
```

or define the list of elements as agnostic e.g. using `items` rather than `datasets`:

```
"collection": {
  "items": {}, # could contain images and labels
}
```

The second option obviously requires  introspection of the metadata to know what the collection is containing.

**Comment by glyg** [comment]
> Not sure what is meant by extend to "tables"?

@will-moore  I meant a collection of tabular data (still with the SMLM use case in mind, you could have the raw positions table, then the filtered or transformed ones) 

I like that `"items"` would be available as a default name, maybe with the constrain that all the items of a collection be of the same type, so inspecting the first item is enough, but would still define `"images"` and `"labels"` as this is the most common use case.

**Comment by tischi** [comment]
In above discussion were some (for me) new things coming up that I did not yet have on my radar:

1. There was the use case that one may only want to load a specific time point (range) or channel (range) from a dataset into a certain view. Maybe, with @will-moore's current proposal that could be realized by a "property" that can _crop_ a dataset?
2. The other new (for me) use case was that, e.g., different time points might be distributed into different datasets and one could combine them on the collection level. I think there one idea was that somehow "tagging" the datasets in collection could be a way. 

Would you say that this correctly describes our discussion (ping @tpietzsch)?

**Comment by tischi** [comment]
I think we had the discussion already somewhere whether this should be called `views` or `collections`. 
For the use case of just loading a specific time point range from an otherwise larger dataset to me the term `view` is more intuitive. In general, I personally like `views` better, as it describes that those are different ways of looking at (opening) the data contained in the ome.zarr container.

**Comment by will-moore** [comment]
I would imagine a collection defines a container (of images and/or labels) that could be viewed or analysed in various ways. e.g.

```
"collection": {
  "images": {
      "image1": {}, "image2": {}, "image3": {}, "image4": {}, "image5": {},
  },
  "labels": {
    "segmentation1": {},
  }
  "views": {
    "my_Z_stack": {
        "type": "concatenate",
        "axis": "z",
        "images": ["image1", "image2", "image3"]
    },
    "my_stitched_view": {
        "type": "stitch",
        "images": ["image4", "image5"]
    },
    "my_segmentation": {
        "type": "overlay",
        "images": ["image4", "image5"]
        "labels": ["segmentation1"]
    },
  }
}
```

In the simplest case, all you need is a `collection` of `images`.

**Comment by tischi** [comment]
@will-moore 

Thanks! That looks very nice! 

What about your very good point that collections/views should be addressable with a unique URL. Could we still make this work in this scenario?

**Comment by will-moore** [comment]
I don't think this gives you a URL to a view, but it does give you a URL to a single collection, which will have to do for now. It's a trade-off: If you have everything distributed at different levels as we do with the current HCS spec, you get URLs to Wells etc, but to get all the info you need for a Plate, you have to traverse every Well (e.g. see https://github.com/hms-dbmi/vizarr/pull/119). 
It *might* work in some cases with `/path/to/collection/?view=my_segmentation`. I seem to remember trying this idea in napari and it didn't work, but maybe that could be addressed.

**Comment by aeisenbarth** [comment]
Should collections be ordered or unordered like sets?

A use case is that images within a collection are usually stacked in a viewer (list of layers), where the order matters. 

Currently, zarr-python's JSON encoder sorts dictionaries alphabetically, so there is some control if the user can compromise on names. Alternatively, order could be specified by something like a `z-index` attribute for each image, but I think a list of images is more intuitive and maintainable than an absolute or relative order attribute.

**Comment by will-moore** [comment]
I was thinking of a collection as being unordered, but could be displayed in order by any client app as it wishes, e.g. like a directory listing or images in an OMERO Dataset.

There were a couple of preferences for Option 2 (dictionary) in https://github.com/ome/ngff/issues/31#issuecomment-791470191
but the `groups` (sub-grouping of images in a collection) are lists in the examples above.

Also, I started to look at how we might view a collection in vizarr.
https://github.com/hms-dbmi/vizarr/pull/124
with data exported from IDR with https://github.com/ome/omero-cli-zarr/pull/88

**Comment by will-moore** [comment]
@aeisenbarth Just re-reading your comment and wanted to clarify: In the case where you're viewing a Z-stack of layers, this would be handled as a single multi-dimensional image, e.g. `x, y, z, channel? time?`.
Images in a collection may have some more complex physical relationships between them, but these would be expressed in some other way e.g. transforms as discussed above and on #28.

**Comment by glyg** [comment]
> If we're going to have a generic "Collection" with anything as items, maybe we should also define a subclass "ImageCollection" with "images":{} or with "items":{} which will be Images?

Sounds very good!

**Comment by will-moore** [comment]
As @joshmoore summarised, we have a choice of these options:

 - 1 ```{ @type: ImageCollection, images: [] }```
 - 2 ```{ @type: ImageCollection, items: [] }```
 - 3 ```{ @type: Collection, images: [] }```
 - 4 ```{ @type: Collection, limitTo: Image, items: [] }```

I prefer 2, but the outstanding question is whether it's possible to have a super-class ```{ @type: Collection, items: [] }``` with any object in `items` AND have a sub-class which restricts `items` to ONLY Image objects?

**Comment by will-moore** [comment]
I don't have a good feel for how to validate JSON-LD. There's a validator at https://validator.schema.org/ but it doesn't seem to fail validation if you use an invalid object e.g. this is valid:
```
{
    "@context": "https://schema.org/",
    "@type": "Invoice",
    "customer":{
      "@type": "PriceSpecification",
      "priceCurrency": "USD"
    }
}
```
although it does fail if you use an Object instead of a Text value.

The playground at https://json-ld.org/playground/ doesn't do validation.
And I don't see any validation support from the Java, Python of JS libs listed at https://json-ld.org/.

@joshmoore mentioned http://shex.io/ which seems to handle validation, but look like you'd need to convert JSON-LD to RDF before validation?

Considering the "bigger picture" of whether we want to use schema to provide cross-language validation of OME-NGFF data, a lot of the rules like "axes list must be equal length to the number of dimensions in the data", aren't going to be amenable to validation via a schema, and will need to be validated by code?
 
So I'm inclined to forget about whether it's possible to use a schema to validate that `items` in `{ @type: ImageCollection, items: [] }` are `Images`.

**Comment by will-moore** [comment]
So, having established that we can use JSON-schema to validate (see https://github.com/ome/ngff/pull/69) I think we need to clarify what we're going with for 1st version of Collections. This is where we got to above:

```
{
  "@type": "ImageCollection",
  "items": [
    {
      "path": "image1",
      "properties": {}
    },
    {"path": "path/to/image2"}
  ]
}
```
So maybe this just needs `"@type": "ImageReference"` or `ImagePath` on each item in the list.
Possibly the `path` key should be `image` as it is a link to an Image object, or is `ImageReference` a subclass of a `ObjectReference` and both should use `path`? @joshmoore?

**Comment by joshmoore** [comment]
A few thoughts but probably mostly just pros/cons to weigh:
 * I'm pretty sure we can apply a default `@type` at all levels but the top. So unless someone is using a non-default type (e.g. `MySpecialImagePath`), we might be able to omit them.
 * We might want to use `@type: "Image"` for references, i.e. something that has been defined elsewhere in the graph. That assumes, though, that we can figure out that a path is a JSON structure that is of type `@Image`.
 * I would tend to not introduce more field names (`{"image": ...}`) to make things more re-usable.

**Comment by will-moore** [comment]
So, like this:

```
{
  "@type": "ImageCollection",
  "items": [
    {
      "@type": "Image",
      "path": "image1",
      "properties": {}
    },
    {
      "@type": "Image",
      "path": "path/to/image2"
    }
  ]
}
```
I'll go with that over at https://github.com/ome/ngff/pull/67/

**Comment by joshmoore** [comment]
Sorry, I realize 'We might want to use @type: "Image" for references,' was unclear. Here with reference I meant "a collection of images":

```
[
  {
    "@type": "Image",
    "@id": "/some/metadata/path"
  }
]
```
where we've omitted fields like `multiscales` with the assumption that the Image "/some/metadata/path" exists somewhere else. If we can make `@id == path` then maybe the two concepts can be shared. If not, then I think what you've written is more of an "ImageRef"

**Comment by will-moore** [comment]
OK, like this:
```
{
  "@type": "ImageCollection",
  "items": [
    {
      "@type": "Image",
      "@id": "image1"
    },
    {
      "@type": "Image",
      "@id": "path/to/image2"
    }
  ]
}
```

And then the Image`.zattrs` itself would have this?
```
{
    "@type": "Image",
    "@id": "image1",     # or "path/to/image2" # or absolute URL ?
    "multiscales": [],
    "omero": {}
}
```
Should the `@id` be the location of the JSON file itself? This will be e.g. "https:/s3.host/path/to/Image2/.zattrs".
Should that match the `@id` in the parent ImageCollection?

**Comment by jkh1** [comment]
Do you intent a collection to be comprised of elements from different sources/datasets or is it intended for grouping/ordering elements inside the same NGFF file? Also I think collections should be distinguished from views. While we could think of views as collections augmented with transformations or relationships for visualization purposes, I think views could also be represented independently as workflows, i.e. a series of transformation steps applied to some input images resulting in an image suitable for visualization.

**Comment by joshmoore** [comment]
Two quick points:
 * The question of "different sources" is a good one and probably overlaps with the [links issue](https://github.com/ome/ngff/issues/13). e.g., if/when we have links, perhaps the collections specification could be agnostic as to the location.
 * The delineation of collection/view (or whatever names may be chosen) is also a good one. I know @StephanPreibisch, @tpietzsch, @StephanPreibisch, @tischi have been working through ideas and the representation as transformation steps was touched on. I do wonder, though, if the more generic "collections" which make up most of the comments above couldn't be a base type that get extended by "views".

**Comment by will-moore** [comment]
As discussed on 7th Jan, we won't try to handle absolute URLs yet. Limit scope for now.
Need to try and see if these options affect the ability to validate via schema (e.g. to check existence of paths), although not sure this will work with remote data on s3?

**Comment by will-moore** [comment]
I can't find any info on JSON-schema validating references within JSON *instances*. The only use of references or ids is within a schema, to refer to other schemas. e.g. https://json-schema.org/understanding-json-schema/structuring.html#ref
Also see
https://github.com/HumanBrainProject/openMINDS_core/issues/256

**Comment by aeisenbarth** [comment]
> I'm not sure I understand this.

I was referring to the proposals above. A path to a Zarr subgroup with images of a single collection does not have the NGFF Collection metadata, so when reading it, it looks like a plain (non-NGFF) Zarr group containing NGFF images. One could check whether the parent directory is a Zarr group and has NGFF Collections metadata, then iterate over all collections to find the one containing these images. This works if I introduce restrictions on my Zarr hierarchy. But for the general case you would have to walk up to the file system root and check every parent because image paths can have several nesting levels.

In short, "a Zarr collection of images" cannot be expressed as a path/URL (but as path to collections + collection name).

**Comment by will-moore** [comment]
At https://github.com/ome/omero-cli-zarr/pull/88 there is code for exporting a Dataset of Images, according to one of the Collection specs discussed above. And there is an example Dataset hosted at https://minio-dev.openmicroscopy.org/idr/v0.3/datasets/idr0043/13901.zarr/.zattrs

That URL defines a collection of Images, listed in the [.zattrs](https://minio-dev.openmicroscopy.org/idr/v0.3/datasets/idr0043/13901.zarr/.zattrs) where we *could* also include the name and other metadata (this example doesn't).

The .zattrs is simple, and we have since decided to use a different structure, but it doesn't look like a plain (non-NGFF) Zarr group because it contains the "collection" dictionary:
```
{
    "collection": {
        "images": {
            "165383_A_1_1.tif": {},
            "165383_A_1_2.tif": {},
            "165383_A_1_4.tif": {},
            ...
        }
    }
}
```
That collection can be viewed in a vizarr PR deployed at
https://deploy-preview-124--vizarr.netlify.app/?source=https://minio-dev.openmicroscopy.org/idr/v0.3/datasets/idr0043/13901.zarr (or see screenshot on the PR above).

Apologies if I'm getting my Zarr terminology wrong, but is that not an example of a path to a Zarr group (or subgroup?) with (limited) Collection metadata and no searching of parent directories needed to find the images?

**Comment by normanrz** [comment]
Copying from image.sc:

We are in the process of integrating OME-NGFF in [webKnossos](https://webknossos.org/). webKnossos has the concept of “Datasets” to organize images. A “Dataset” consists of multiple multi-scale layers, which can be variety of images including raw EM data, multi-channel images, instance segmentations, ML probability maps etc. All layers in a Dataset have a shared coordinate system and somewhat belong together in a semantic sense.

With regards to this proposal, I was wondering if there should be different types of collections (those with shared coordinate spaces and ones with separate spaces)?

**Comment by tischi** [comment]
@normanrz @joshmoore @will-moore 
1. Should we consider meeting to work on this?
2. I wonder whether such a collection spec is OME-Zarr specific or whether it could be on top of it in a sense that also a few important other image data formats could be handled?

**Comment by normanrz** [comment]
> 1. Should we consider meeting to work on this?

Yes!



> 2\. I wonder whether such a collection spec is OME-Zarr specific or whether it could be on top of it in a sense that also a few important other image data formats could be handled?

I think it fits nicely in the OME-Zarr world. For our purposes, there is no requirement to include other image formats.

**Comment by tischi** [comment]
For me, having a collection spec for a bunch of TIFF files also would be very handy.
We can see whether the collection spec will need to refer to OME-Zarr specifics. 
So far all the JSON snippets that have posted above seem to also work for other image files...

**Comment by jluethi** [comment]
I'd also be very interested to see where the collection work is going! We're just interested in OME-Zarr images and use the HCS spec heavily. But would be interesting to hear how this can generalize to other collections and how metadata about the contents of the collection is represented :)

**Comment by will-moore** [comment]
I think we got bogged down previously because we tried to do too much, so let's see if we can list the requirements that we all agree on before we try to find solutions. Here's a possible list, and I'm only selecting 2 items to start. Maybe others could copy this list and select items or add their own items?

 - [x] Simple `collection` that lists child images: e.g. https://github.com/ome/ngff/issues/31#issuecomment-1011869652
 - [x] A way to store extra metadata on the collection: e.g. `properties: {}`
 - [ ] Image in multiple collections (needs absolute paths or `../../relative/path/to/sibling/collection`)
 - [ ] Replacement for Plate: e.g. https://github.com/ome/ngff/issues/31#issuecomment-802064440
 - [ ] Types, e.g. `"@type": "ImageCollection"` subclassing `Collection`etc e.g. https://github.com/ome/ngff/issues/31#issuecomment-982629030 questions remain
 - [ ] Different collection types/metadata for shared coordinates spaces (views) and separate spaces
 - [ ] Multiple collections in a single file? e.g. https://github.com/ome/ngff/issues/31#issuecomment-792582677
 - [ ] Collections can contain collections?

**Comment by tischi** [comment]
How do I copy the list? :-) If someone tells me I am happy to reformat what I write here:

For my current work I (and I think @jluethi may be interested) would need one more thing:

Some way to **specify the shape and pixel type metadata for a collection of images**, such that I don't need to open all of them during initialisation (use case is similar to HCS where we have hundreds of images with identical metadata that we want to open in a "grid view"); essentially I want to be able to express that "here is a collection of images and they all have the same shape (dimensions) and pixel type, namely ... 
I realise now that this may be covered by: [X] A way to store extra metadata on the collection`. 

- [X] Attach rendering metadata to image in the collection
   - For me it would be important to express that _all_ images in the collection should have the _same_ rendering settings (same use-case as above).

**Comment by jluethi** [comment]
> Some way to specify the shape and pixel type metadata for a collection of images, such that I don't need to open all of them during initialisation (use case is similar to HCS where we have hundreds of images with identical metadata that we want to open in a "grid view")

> it would be important to express that all images in the collection should have the same rendering settings (same use-case as above).

In the use case where we assume all items of the collection (e.g. all wells) have the same metadata, I don't see a huge difference about whether I load the metadata from the collection level or from a single (randomly chosen? alphabetically first?) image. A simple flag may be sufficient then.
It gets most interesting to us when we know that the collection items (e.g. different different wells) may have images of different shapes. This, we'd need to know upon plate loading (to initialize the pyramids correctly) and loading of metadata from hundreds of images becomes a bottleneck (see discussion in this PR to add support to ome-zarr-py to load HCS plates with varying well sizes: https://github.com/ome/ome-zarr-py/pull/241)

**Comment by aeisenbarth** [comment]
We had a need for collections and did a most minimal implementation of [this](https://github.com/ome/ngff/issues/31#issuecomment-792582677) proposal. We were using it for grouping images associated with each well (collection A1 → {modality1, modality2…}). Now with the plate spec not being marked transitional, we are contemplating moving to that (storing each modality as a separate plate), because it is more standards-compliant than our own extension, we practically don't encounter plate layouts that are not row/column based, and we don't need assumptions that each collection actually contains all expected images.

Things to consider:
- Do you want to be able to reference a collection like a file path?
- It is practical that images are still self-contained and complete. When attaching metadata at the collection level, it should probably only be a summary/copy of what is at the image level, to avoid that when reading a child in the hierarchy you have to look up its parent (if it has).
- Maybe most flexible is a design where the optional image properties follow the NGFF image spec. That way you could have overrides, let's say when loading image1 as part of this collection, use a different coordinate transformation.

@jluethi, couldn't you define (sub)collections based on same metadata? Unless the partitioning into subcollections is different for every metadata property.

**Comment by will-moore** [comment]
@tischi here's the markdown for the list to copy/paste

```
 - [x] Simple `collection` that lists child images: e.g. https://github.com/ome/ngff/issues/31#issuecomment-1011869652
 - [x] A way to store extra metadata on the collection: e.g. `properties: {}`
 - [ ] Image in multiple collections (needs absolute paths or `../../relative/path/to/sibling/collection`)
 - [ ] Replacement for Plate: e.g. https://github.com/ome/ngff/issues/31#issuecomment-802064440
 - [ ] Types, e.g. `"@type": "ImageCollection"` subclassing `Collection`etc e.g. https://github.com/ome/ngff/issues/31#issuecomment-982629030 questions remain
 - [ ] Different collection types/metadata for shared coordinates spaces (views) and separate spaces
 - [ ] Multiple collections in a single file? e.g. https://github.com/ome/ngff/issues/31#issuecomment-792582677
 - [ ] Collections can contain collections? 
 ```

**Comment by tischi** [comment]
Use-case wise I realised that for me it would also be interesting to know whether
- all images are 2D
- all images have the same spatial extends (but may have different number of time points)

I wonder now, if this collection.json is mainly meant to be read and written by a computer, whether it would be easier and more generic to simply have the option to add for to each image all sorts of metadata from the .zarray files, such as dtype, shape, multiscales. Since this is in a single file, for a computer it would still be very fast to figure out whether "all are 2D"  or "all have same shape" a.s.o.. I think the same argument can be made for rendering settings. 

What about the following: 

We first figure out all `properties` that we would like to be able to add to each individual image of the collection?

Then, as second step, we discuss whether it is worth to have a `common_properties` field where we can say things that apply to all images?

**Comment by normanrz** [comment]
> @normanrz Do you need to specify rendering settings for each image (rather than a rendering for the whole collection)? Is this just to save time, to avoid loading the settings from each `image.zarr/.zattrs`?

This came up in the community call: Think of a collection as the definition of the view state in a visualization tool. Now, the render settings feel better placed in the view than with the individual images because you may want to have the same image in different views with different settings. For example, you may want to choose different colors for an image in different views. `visible` also only really makes sense in the context of a view. So to clarify my proposal: move all rendering settings from the image into the collection.

> Making these "image properties follow the NGFF image spec" as @aeisenbarth suggested makes sense when possible (e.g. name and rendering settings that are in the NGFF multiscales metadata), but might be tricky for others that are in the Zarr .zarray data (dtype, shape).

I want to mention that idea of consolidated metadata has come up on the Zarr-level quite a few times. This would mean that the `zarr.json` of a group could store the `zarr.json` contents of descendent groups and arrays. This is primarily a performance optimization to save IO.
Even though there is no formal proposal for this yet, I don't think we should reinvent that on the OME-level. At least for core metadata (shape, dtype etc). The OME-Zarr metadata would still be duplicated for each image in the consolidated json. I don't think that is a big problem, because I assume for 100s of images it should still be on the order of 1MB. Also, the json probably compresses very well.

Storing information like 

> {
>   "allSameShape": true.
>   "allSameDataType": true,
>   "allSameRendering": true,
> }

or

> * all images are 2D
> * all images have the same spatial extends (but may have different number of time points)

seems like redundant aggregations that could easily be derived by the visualization or processing software (given consolidated metadata).

**Comment by moi90** [comment]
> So far, the only collections currently in the spec are [bioformats2raw.layout](https://ngff.openmicroscopy.org/0.5/index.html#bf2raw) and [plates/wells/images](https://ngff.openmicroscopy.org/0.5/index.html#plate-md).

(~ [Will Moore](https://forum.image.sc/u/will-moore) / OME Team: https://forum.image.sc/t/save-a-single-labels-dataset-into-an-ome-zarr/93505/42)

**Comment by jni** [comment]
It's probably worth adding a link to this issue in the [RFC-8 references section](https://hackmd.io/rQPfIMw0Q9SqP9WWspCiLQ#Prior-art-and-references) @normanrz @jluethi
